app/waf/src/services/bigip.service.ts

```diff
@@ -205,6 +213,98 @@ export class BigIpManager {
     return resObj;
   }
 
+  async getDOStatus(): Promise<string> {
+    await this.mustBeReachable();
+
+    let url = `${this.baseUrl}/mgmt/shared/declarative-onboarding/info`;
+    let response = await this.bigipService.getInfo(url, this.cred64Encoded);
+    let resObj = JSON.stringify(response);
+    return resObj;
+  }
+
+  async uploadDO(): Promise<string> {
+    // if the local file doesn't exist, throw execption.
+    //read the file's contant into a buf and calculate its length
+    //call the bigipService.uploadDO to upload the RPM to Bigip
+    //possibly check  if the upload succeeds or not
+    //await this.mustBeReachable();
+    const filename = '/var/tmp/DO/F5_DO_RPM_PACKAGE.rpm';
+    let fs = require('fs');
+    if (!fs.existsSync(filename)) {
+      console.log("DO RPM file doesn't exist.");
+      throw new Error("DO RPM file doesn't exist.");
+    }
+    let fstats = fs.statSync(filename);
+    try {
+      let url = `${this.baseUrl}/mgmt/shared/file-transfer/uploads/F5_DO_RPM_PACKAGE.rpm`;
+      let buffer = fs.readFileSync(filename, {endcoding: 'utf8'});
+      buffer.toString('utf8');
+      let response = await this.bigipService.uploadDO(
+        url,
+        this.cred64Encoded,
+        fstats.size - 1,
+        fstats.size,
+        buffer,
+      );
+      let resObj = JSON.stringify(response);
+      return resObj;
+    } catch {
```

Comment:
it's better to `catch(error) {throw new Error(Upload DO RPM file error.: ${error});}`, so that we can know better why. 

Another better option is not catch it, and make it throw to upper caller(let upper caller to handle it.)

Reply:
Sure, will make the change accordingly.

Thanks